IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,660 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <Foundation/Foundation.h>
+#import "MSIDAccountCacheItem.h"
+#import "MSIDAccountItemSerializer.h"
+#import "MSIDAccountType.h"
+#import "MSIDCredentialCacheItem.h"
+#import "MSIDCredentialItemSerializer.h"
+#import "MSIDCredentialType.h"
+#import "MSIDError.h"
+#import "MSIDJsonSerializer.h"
+#import "MSIDLogger+Internal.h"
+#import "MSIDLogger.h"
+#import "MSIDMacKeychainTokenCache.h"
+#import "MSIDUserInformation.h"
+#import "NSString+MSIDExtensions.h"
+#import "MSIDKeychainUtil.h"
+#import "MSIDMacKeychainAccountCacheKey.h"
+
+/**
+This Mac cache stores serialized account and credential objects in the macOS "login" Keychain.
+There are three types of items stored:
+  1) Secret shareable artifacts (SSO credentials: Refresh tokens, other global credentials)
+  2) Non-secret shareable artifacts (account metadata)
+  3) Secret non-shareable artifacts (access tokens, ID tokens)
+
+In addition to the basic account & credential properties, the following definitions are used below:
+  <account_id>           :  “<home_account_id>-<environment>”
+  <credential_id>        : “<credential_type>-<client_id>-<realm>”
+  <access_group>         : e.g. "com.microsoft.officecache"
+  <authority_account_id> : e.g. "joe@contoso.com"
+
+Below, attributes marked with "*" are primary keys for the keychain.
+For password items, the primary attributes are kSecAttrAccount and kSecAttrService.
+Other secondary attributes do not make items unique, only the primary attributes.
+
+Type 1 (Secret shareable artifacts) Keychain Item Attributes
+============================================================
+ATTRIBUTE         VALUE
+~~~~~~~~~         ~~~~~~~~~~~~~~~~~~~~~~~~
+*kSecClass        kSecClassGenericPassword
+*kSecAttrAccount  <access_group>
+*kSecAttrService  “Microsoft Credentials”
+kSecValueData     JSON data (UTF8 encoded) – shared credentials (multiple credentials saved in one keychain item)
+
+Type 1 JSON Data Example:
+{
+  "cache": {
+    "credentials": {
+      "account_id1": {
+        "credential_id1": "credential1 payload",
+        "credential_id2": "credential2 payload"
+      },
+      "account_id2": {
+        "credential_id1": "credential1 payload",
+        "credential_id2": "credential2 payload"
+      }
+    }
+  }
+}
+
+Type 2 (Non-secret shareable artifacts) Keychain Item Attributes
+================================================================
+ATTRIBUTE         VALUE
+~~~~~~~~~         ~~~~~~~~~~~~~~~~~~~~~~~~
+*kSecClass        kSecClassGenericPassword
+*kSecAttrAccount  <access_group>-<account_id>
+*kSecAttrService  <realm>
+kSecAttrGeneric   <authority_account_id>
```

Comment:
this is outdated.

Reply:
<authority_account_id> changed to <username>.